src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.DecryptTokenResult.cs

```diff
@@ -0,0 +1,151 @@
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            bool decryptionSucceeded = false;
+            bool algorithmNotSupportedByCryptoProvider = false;
+            byte[] decryptedTokenBytes = null;
+
+            // keep track of exceptions thrown, keys that were tried
+            StringBuilder exceptionStrings = null;
+            StringBuilder keysAttempted = null;
+            string zipAlgorithm = null;
+            foreach (SecurityKey key in decryptionParameters.Keys)
+            {
+                var cryptoProviderFactory = validationParameters.CryptoProviderFactory ?? key.CryptoProviderFactory;
+                if (cryptoProviderFactory == null)
+                {
+                    if (LogHelper.IsEnabled(EventLogLevel.Warning))
+                        LogHelper.LogWarning(TokenLogMessages.IDX10607, key);
+
+                    continue;
+                }
+
+                try
+                {
+                    if (!cryptoProviderFactory.IsSupportedAlgorithm(jsonWebToken.Enc, key))
+                    {
+                        if (LogHelper.IsEnabled(EventLogLevel.Warning))
+                            LogHelper.LogWarning(TokenLogMessages.IDX10611, LogHelper.MarkAsNonPII(decryptionParameters.Enc), key);
+
+                        algorithmNotSupportedByCryptoProvider = true;
+                        continue;
+                    }
+
+                    AlgorithmValidationResult result = validationParameters.AlgorithmValidator(zipAlgorithm, key, jsonWebToken, validationParameters, callContext);
+                    if (!result.IsValid)
+                    {
+                        (exceptionStrings ??= new StringBuilder()).AppendLine(result.ExceptionDetail.MessageDetail.Message);
+                        continue;
+                    }
+
+                    decryptedTokenBytes = DecryptToken(
+                        cryptoProviderFactory,
+                        key,
+                        jsonWebToken.Enc,
+                        jsonWebToken.CipherTextBytes,
+                        jsonWebToken.HeaderAsciiBytes,
+                        jsonWebToken.InitializationVectorBytes,
+                        jsonWebToken.AuthenticationTagBytes);
+
+                    zipAlgorithm = jsonWebToken.Zip;
+                    decryptionSucceeded = true;
+                    break;
+                }
+#pragma warning disable CA1031 // Do not catch general exception types
+                catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                {
+                    (exceptionStrings ??= new StringBuilder()).AppendLine(ex.ToString());
+                }
+
+                if (key != null)
+                    (keysAttempted ??= new StringBuilder()).AppendLine(key.ToString());
+            }
+
+            ExceptionDetail exceptionDetail = ValidateDecryption(
```

Comment:
I would recommend consider removing this call as after refactoring into a method that has one key would simplify the logic where this is not needed.

Reply:
Adding a work item to look at this after we have our `ValidateTokenAsync` implementation working.